.ci/magician/cmd/collect_nightly_test_status.go

```diff
@@ -0,0 +1,213 @@
+
+
+	The command expects the following argument(s):
+	1. Custom test date in YYYY-MM-DD format. default: ""(current time when the job is executed)
+
+	It then performs the following operations:
+	1. Collects nightly test status of the execution day or the specified test date (if provided)
+	2. Stores the collected data in JSON files
+	3. Uploads the JSON files to GCS
+
+	The following environment variables are required:
+` + listCNTSRequiredEnvironmentVariables(),
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		env := make(map[string]string)
+		for _, ev := range cntsRequiredEnvironmentVariables {
+			val, ok := os.LookupEnv(ev)
+			if !ok {
+				return fmt.Errorf("did not provide %s environment variable", ev)
+			}
+			env[ev] = val
+		}
+
+		tc := teamcity.NewClient(env["TEAMCITY_TOKEN"])
+		gcs := cloudstorage.NewClient()
+
+		date := time.Now()
+		customDate := args[0]
+		//
+		if customDate != "" {
+			parsedDate, err := time.Parse("2006-01-02", customDate) // input format YYYY-MM-DD
+			// Set the time to 6pm PT
+			date = time.Date(parsedDate.Year(), parsedDate.Month(), parsedDate.Day(), 18, 0, 0, 0, parsedDate.Location())
+			if err != nil {
+				return fmt.Errorf("invalid input time format: %w", err)
+			}
+		}
+
+		return execCollectNightlyTestStatus(date, tc, gcs)
+	},
+}
+
+func listCNTSRequiredEnvironmentVariables() string {
+	var result string
+	for i, ev := range cntsRequiredEnvironmentVariables {
+		result += fmt.Sprintf("\t%2d. %s\n", i+1, ev)
+	}
+	return result
+}
+
+func execCollectNightlyTestStatus(now time.Time, tc TeamcityClient, gcs CloudstorageClient) error {
+	lastday := now.AddDate(0, 0, -1)
+	formattedStartCut := lastday.Format(time.RFC3339)
+	formattedFinishCut := now.Format(time.RFC3339)
+	date := now.Format("2006-01-02")
+
+	err := createTestReport(provider.GA, tc, gcs, formattedStartCut, formattedFinishCut, date)
+	if err != nil {
+		return fmt.Errorf("Error getting GA nightly test status: %w", err)
+	}
+
+	err = createTestReport(provider.Beta, tc, gcs, formattedStartCut, formattedFinishCut, date)
+	if err != nil {
+		return fmt.Errorf("Error getting Beta nightly test status: %w", err)
+	}
+
+	return nil
+}
+
+func createTestReport(pVersion provider.Version, tc TeamcityClient, gcs CloudstorageClient, formattedStartCut, formattedFinishCut, date string) error {
+	// Get all service test builds
+	builds, err := tc.GetBuilds(pVersion.TeamCityNightlyProjectName(), formattedFinishCut, formattedStartCut)
+	if err != nil {
+		return err
+	}
+
+	var testInfoList []TestInfo
+	for _, build := range builds.Builds {
+		serviceTestResults, err := tc.GetTestResults(build)
+		if err != nil {
+			return fmt.Errorf("failed to get test results: %v", err)
+		}
+		if len(serviceTestResults.TestResults) == 0 {
+			continue
+		}
+		// Get service package name
+		serviceName, err := convertServiceName(build.BuildTypeId)
```

Comment:
It looks like this will result in SERVICE_SWEEPER being counted as a package called SWEEPER... it might be nice to special case that.

Reply:
yeah, it makes sense. Currently, sweeper is skipped reply on the previous check `len(serviceTestResults.TestResults) == 0` but after the refactor of sweepers, it may not be working anymore. I agree it's nice to explicit case it.

Reply:
Added an explicit check for sweeper